### DISCUSSION TITLE
Madmin and Shrine (and its fake attributes)

### DISCUSSION THREAD

**Discussion by johlym** [excid3/madmin#124] [Q&A]
I'm not sure if I'm even correctly understanding what I'm seeing. Long and short is I'm attempting to implement [Shrine](https://github.com/shrinerb/shrine). All is well until I get to the stage where I have to tell Madmin that I need a hidden field and real field to appear. Their example is simple:

```
form_for @photo do |f|
  f.hidden_field :image, value: @photo.cached_image_data
  f.file_field :image
  f.submit
end
```

Seems simple enough for sure, but the field Shrine wants defined doesn't actually have a dedicated attribute in the model... it's fake... or as Shrine calls it, `virtual`. 

This poses a problem when Madmin tries to render the form and gives me something to the effect of:

```
Madmin couldn't find attribute or association 'image' on Ad model.
```

I do not doubt it couldn't find it, because it's not real. Any ideas on how I can get around this? I had to shelve ActiveStorage for this project because it lacked public URL support without lots of hacks. I'm thinking I might have unknowingly traded one set of hacks for another.

**Comment by excid3** [comment]
You'll want to specify the field type manually for now. `field :image, field: Fields::Attachment`. This is mainly for ActiveStorage attachments, but it might need a custom field since Attachment may not render the correct URL since it functions differently.

If someone wants, we could make a PR that:
- Modifies `infer_type` in Resource to check for a Shrine value from the method
- Either update `Field::Attachment` to also support Shrine uploads or add a `Field::Image` type that knows how to render the URL

**Comment by johlym** [comment]
Thanks for replying so quickly! Maybe I'm not grokking the proper way to make this happen. setting:

```
class AdResource < Madmin::Resource
  ...
  attribute :image, field: Fields::Attachment
  ...
end
```

I get 

```
uninitialized constant AdResource::Fields
```

Am I misinterpreting your workaround?
